cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java

```diff
@@ -133,57 +136,193 @@ public void apply() throws Exception {
+                                             STOP_STATUSES.get((i + 1) % STOP_STATUSES.size())));
+    }
+    // STARTING status will be ignored if there's any existing record
+    assertPersistedStatus(metadataStoreDataset, txnl, ProgramRunStatus.SUSPENDED,
+                          ImmutableList.of(ProgramRunStatus.STARTING, ProgramRunStatus.SUSPENDED,
+                                           ProgramRunStatus.STARTING));
+    assertPersistedStatus(metadataStoreDataset, txnl, ProgramRunStatus.RUNNING,
+                          ImmutableList.of(ProgramRunStatus.STARTING, ProgramRunStatus.RUNNING,
+                                           ProgramRunStatus.STARTING));
+  }
+
+  private void assertPersistedStatus(final AppMetadataStore metadataStoreDataset, TransactionExecutor txnl,
+                                     final long startSourceId, final long runningSourceId,
+                                     final long killedSourceId, ProgramRunStatus expectedRunStatus) throws Exception {
     // Add some run records
-    TreeSet<Long> expected = new TreeSet<>();
     ApplicationId application = NamespaceId.DEFAULT.app("app");
     final ProgramId program = application.program(ProgramType.WORKFLOW, "program");
-    final RunId runId = RunIds.generate(10000);
-    // Program completed status is not recorded with a smaller sourceId
+    // use sourceId's to ensure runId is unique for each run
+    final RunId runId = RunIds.generate(10000 + startSourceId + runningSourceId + killedSourceId);
+    final AtomicReference<RunRecordMeta> resultRecord = new AtomicReference<>();
     txnl.execute(new TransactionExecutor.Subroutine() {
       @Override
       public void apply() throws Exception {
         metadataStoreDataset.recordProgramStart(program, runId.getId(), RunIds.getTime(runId, TimeUnit.SECONDS),
                                                 null, ImmutableMap.<String, String>of(),
                                                 ImmutableMap.<String, String>of(),
-                                                AppFabricTestHelper.createSourceId(10000));
+                                                AppFabricTestHelper.createSourceId(startSourceId));
+        metadataStoreDataset.recordProgramRunning(program, runId.getId(), RunIds.getTime(runId, TimeUnit.SECONDS),
+                                                  null, AppFabricTestHelper.createSourceId(runningSourceId));
         metadataStoreDataset.recordProgramStop(program, runId.getId(), RunIds.getTime(runId, TimeUnit.SECONDS),
                                                ProgramRunStatus.KILLED,
-                                               null, AppFabricTestHelper.createSourceId(10));
+                                               null, AppFabricTestHelper.createSourceId(killedSourceId));
+        resultRecord.set(metadataStoreDataset.getRun(program, runId.getId()));
       }
     });
+    Assert.assertEquals(expectedRunStatus, resultRecord.get().getStatus());
```

Comment:
if you move this assert into the transaction, you don't need the AtomicReference